### DISCUSSION TITLE
GIS  support

### DISCUSSION THREAD

**Discussion by v0y4g3r** [GreptimeTeam/greptimedb#2187] [Ideas]
In many cases, time series data appears along with GIS data, like GPS coordinates. It's a big plus to support GIS data in GreptimeDB. This feature can be divided into serveral parts:
- Supporting GIS data types, like `Points`, `Polygon`
- Supporting spatial index, such as R-tree/KD-tree
- Supporting spatial functions, such as [nearest-neighbour search](https://postgis.net/workshops/postgis-intro/knn.html) and geofencing
- Integrates with other GIS systems

Some Rust crates that maybe useful:
- [GeoRust](https://github.com/georust)

**Comment by sunng87** [comment]
Adding some information for those who are interested in this:

- GIS primitives are defined in https://github.com/georust/geo/tree/main/geo-types we can reuse these types. `Point` is at highest priority, followed by `LineString` and `Polygon`.
- We will need to investigate how to represent geo types in Arrow memory format. Previous work has been discussed in geopandas project (as pandas 2.0 uses Arrow backend), and more specifically in [geoarrow](https://github.com/geoarrow/geoarrow) repo. 
- A typical usage of geo data in time-series context is trajectory data and analytical function, we can learn from [Ganos](https://www.vldb.org/pvldb/vol15/p3483-chen.pdf), [MobilityDB](https://github.com/MobilityDB/MobilityDB). Trajectory analysis functions can be useful for these spatial-temporal scenarios:
  - pointAtTime
  - timeAtPoint
  - velocityAtTime
  - acclerationAtTime

**Comment by lewiszlw** [comment]
[geoarrow](https://github.com/geoarrow/geoarrow-rs) has good support for concrete types(e.g. PointArray) but not for mixed type (e.g. MixedGeometryArray). And it has integrated many useful algorithms from [georust/geo](https://github.com/georust/geo). Very convenient to use this crate to extend Datafusion.
